File: backend-java/src/main/java/com/groceriesapp/service/ShoppingListService.java
package com.groceriesapp.service;

import com.groceriesapp.model.Household;
import com.groceriesapp.model.ShoppingList;
import com.groceriesapp.model.ShoppingListItem;
import com.groceriesapp.repository.HouseholdRepository;
import com.groceriesapp.repository.ShoppingListItemRepository;
import com.groceriesapp.repository.ShoppingListRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ShoppingListService {
    
    private final ShoppingListRepository shoppingListRepository;
    private final ShoppingListItemRepository shoppingListItemRepository;
    private final HouseholdRepository householdRepository;
    
    @Transactional
    public ShoppingList getOrCreateShoppingList(Long householdId) {
        return shoppingListRepository.findByHouseholdId(householdId)
            .orElseGet(() -> {
                Household household = householdRepository.findById(householdId)
                    .orElseThrow(() -> new RuntimeException("Household not found"));
                
                ShoppingList shoppingList = new ShoppingList();
                shoppingList.setName("Default Shopping List");
                shoppingList.setHousehold(household);
                return shoppingListRepository.save(shoppingList);
            });
    }
    
    @Transactional
    public ShoppingListItem addItemToList(Long listId, String name, Integer quantity) {
        ShoppingList shoppingList = shoppingListRepository.findById(listId)
            .orElseThrow(() -> new RuntimeException("Shopping list not found"));
        
        ShoppingListItem item = new ShoppingListItem();
        item.setName(name);
        item.setQuantity(quantity);
        item.setIsPurchased(false);
        item.setShoppingList(shoppingList);
        
        return shoppingListItemRepository.save(item);
    }
    
    @Transactional
    public ShoppingListItem toggleItemPurchased(Long itemId) {
        ShoppingListItem item = shoppingListItemRepository.findById(itemId)
            .orElseThrow(() -> new RuntimeException("Shopping list item not found"));
        
        item.setIsPurchased(!item.getIsPurchased());
        return shoppingListItemRepository.save(item);
    }
    
    /**
     * Placeholder for smart shopping suggestions based on consumption patterns.
     */
    public List<String> getSmartSuggestions(Long householdId) {
        // In a real implementation, this would analyze consumption history
        return Arrays.asList("Milk", "Eggs", "Bread", "Apples", "Chicken");
    }
}

File: backend/src/main/kotlin/com/groceriesapp/service/ShoppingListService.kt
package com.groceriesapp.service

import com.groceriesapp.model.Household
import com.groceriesapp.model.ShoppingList
import com.groceriesapp.model.ShoppingListItem
import com.groceriesapp.repository.ShoppingListRepository
import com.groceriesapp.repository.ShoppingListItemRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ShoppingListService(
    private val shoppingListRepository: ShoppingListRepository,
    private val shoppingListItemRepository: ShoppingListItemRepository
) {

    @Transactional
    fun getOrCreateShoppingList(household: Household): ShoppingList {
        return shoppingListRepository.findByHouseholdId(household.id) ?: createShoppingList(household)
    }

    @Transactional
    fun createShoppingList(household: Household): ShoppingList {
        val newList = ShoppingList(name = "Main Shopping List", household = household)
        return shoppingListRepository.save(newList)
    }

    @Transactional
    fun addItemToList(list: ShoppingList, name: String, quantity: Int): ShoppingListItem {
        val newItem = ShoppingListItem(name = name, quantity = quantity, shoppingList = list)
        list.items.add(newItem)
        return shoppingListItemRepository.save(newItem)
    }

    @Transactional
    fun togglePurchased(itemId: Long): ShoppingListItem {
        val item = shoppingListItemRepository.findById(itemId).orElseThrow {
            NoSuchElementException("Shopping list item with ID $itemId not found.")
        }
        val updatedItem = item.copy(isPurchased = !item.isPurchased)
        return shoppingListItemRepository.save(updatedItem)
    }

    /**
     * Simulates a smart suggestion logic based on low stock/consumed items.
     * In a real app, this would analyze consumption history and current inventory.
     */
    fun getSmartSuggestions(household: Household): List<String> {
        // Placeholder logic: Suggest common items that might be low
        val currentItems = listOf("Milk", "Eggs", "Bread")
        val suggestions = mutableListOf<String>()

        if (!currentItems.contains("Milk")) suggestions.add("Milk")
        if (!currentItems.contains("Eggs")) suggestions.add("Eggs")
        if (!currentItems.contains("Bread")) suggestions.add("Bread")
        if (suggestions.isEmpty()) suggestions.add("Coffee")

        return suggestions
    }
}
